### ISSUE TITLE
Error extracting protoc for version 3.5.1

### ISSUE THREAD

**Issue by cjjnjust** [os72/protoc-jar-maven-plugin#68]
When I building [Apache Parquet](https://github.com/apache/parquet-mr) (commit: e9c283728cfd634677975affa367e13c4bef6c2a)  with the plugin version (3.6.0.1) on ubuntu 18.04, I met following error:

[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.6.0.1:run (generate-sources) on project parquet-protobuf: Error extracting protoc for version 3.5.1: Unsupported platform: protoc-3.5.1-linux-x86_64.exe -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.6.0.1:run (generate-sources) on project parquet-protobuf: Error extracting protoc for version 3.5.1
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error extracting protoc for version 3.5.1
    at com.github.os72.protocjar.maven.ProtocJarMojo.performProtoCompilation (ProtocJarMojo.java:361)
    at com.github.os72.protocjar.maven.ProtocJarMojo.execute (ProtocJarMojo.java:327)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: java.io.FileNotFoundException: Unsupported platform: protoc-3.5.1-linux-x86_64.exe
    at com.github.os72.protocjar.Protoc.extractProtoc (Protoc.java:231)
    at com.github.os72.protocjar.Protoc.extractProtoc (Protoc.java:192)
    at com.github.os72.protocjar.Protoc.extractProtoc (Protoc.java:188)
    at com.github.os72.protocjar.maven.ProtocJarMojo.performProtoCompilation (ProtocJarMojo.java:347)
    at com.github.os72.protocjar.maven.ProtocJarMojo.execute (ProtocJarMojo.java:327)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:134)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:51)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:309)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:194)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:107)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:955)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

**Comment by os72** [comment]
Could you also show the log messages from stdout (`protoc-jar: ...`)

Also, what do these output:
`java -jar protoc-jar-maven-plugin-3.6.0.1.jar -pp`
`java -jar protoc-jar-maven-plugin-3.6.0.1.jar --version`
`java -jar protoc-jar-maven-plugin-3.6.0.1.jar --version -v351`

**Comment by os72** [comment]
It looks like it cannot access maven central (to download 3.5.1)
Are you using a maven mirror?

Version 3.6.0.2 does a better job logging about this
https://github.com/os72/protoc-jar-maven-plugin/blob/master/RELEASE_NOTES.md
https://github.com/os72/protoc-jar/issues/56

For now, using `protocArtifact` instead of `protocVersion` may fix it for you
`<protocArtifact>com.google.protobuf:protoc:3.5.1</protocArtifact>`

**Comment by cjjnjust** [comment]
You are right, maven works after a proxy,  the following output from is maven -X.
```
      url: https://repo.maven.apache.org/maven2/
   layout: default
    proxy: dev-proxy.oa.com:8080
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]
,       id: jitpack.io
      url: https://jitpack.io
   layout: default
    proxy: dev-proxy.oa.com:8080
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]
,       id: apache.snapshots
      url: http://repository.apache.org/snapshots
   layout: default
    proxy: dev-proxy.oa.com:8080
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]
,       id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
    proxy: dev-proxy.oa.com:8080
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
]

```

The following is output when I use 3.6.0.2.
```
$ java -jar ~/.m2/repository/com/github/os72/protoc-jar-maven-plugin/3.6.0.2/protoc-jar-maven-plugin-3.6.0.2.jar -pp
------------------------------------------------------------------------
Detecting the operating system and CPU architecture
------------------------------------------------------------------------
os.detected.name: linux
os.detected.arch: x86_64
os.detected.version: 4.15
os.detected.version.major: 4
os.detected.version.minor: 15
os.detected.release: ubuntu
os.detected.release.version: 18.04
os.detected.release.like.ubuntu: true
os.detected.release.like.debian: true
os.detected.classifier: linux-x86_64
Detected platform: linux-x86_64 (linux/amd64)
$ java -jar ~/.m2/repository/com/github/os72/protoc-jar-maven-plugin/3.6.0.2/protoc-jar-maven-plugin-3.6.0.2.jar --version -v351
protoc-jar: protoc version: 3.5.1, detected platform: linux-x86_64 (linux/amd64)
protoc-jar: downloading: http://central.maven.org/maven2/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: http://central.maven.org/maven2/com/github/os72/protoc/maven-metadata.xml
protoc-jar: downloading: http://central.maven.org/maven2/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: http://central.maven.org/maven2/com/github/os72/protoc/maven-metadata.xml
protoc-jar: java.io.FileNotFoundException: Unsupported platform: protoc-3.5.1-linux-x86_64.exe
java.io.FileNotFoundException: Unsupported platform: protoc-3.5.1-linux-x86_64.exe
	at com.github.os72.protocjar.Protoc.extractProtoc(Protoc.java:231)
	at com.github.os72.protocjar.Protoc.extractProtoc(Protoc.java:192)
	at com.github.os72.protocjar.Protoc.runProtoc(Protoc.java:76)
	at com.github.os72.protocjar.Protoc.runProtoc(Protoc.java:63)
	at com.github.os72.protocjar.Protoc.main(Protoc.java:53)
$ java -jar ~/.m2/repository/com/github/os72/protoc-jar-maven-plugin/3.6.0.2/protoc-jar-maven-plugin-3.6.0.2.jar --version
protoc-jar: protoc version: 3.6.0, detected platform: linux-x86_64 (linux/amd64)
protoc-jar: embedded: bin/3.6.0/protoc-3.6.0-linux-x86_64.exe
protoc-jar: executing: [/tmp/protocjar15252739869151241796/bin/protoc.exe, --version]
libprotoc 3.6.0
```
The version 3.5.1 works fine since it embedded the one
`
protoc-jar: protoc version: 3.5.1, detected platform: linux-x86_64 (linux/amd64)
protoc-jar: embedded: bin/3.5.1/protoc-3.5.1-linux-x86_64.exe
protoc-jar: executing: [/tmp/protocjar8692605776304732429/bin/protoc.exe, --version]
libprotoc 3.5.1
`

So the question is why proxy settings not working in this jar? Does it read proxy settings from maven settings.xml? Or it has to specify separately?

**Comment by os72** [comment]
That's because the download is performed inside `protoc-jar` lib which is not a maven plugin. There is an enhancement request
https://github.com/os72/protoc-jar/issues/57

But you can configure the plugin to force maven download/resolution method
`<protocArtifact>com.google.protobuf:protoc:3.5.1</protocArtifact>`

**Comment by os72** [comment]
https://issues.apache.org/jira/browse/PARQUET-1493

**Comment by os72** [comment]
@cjjnjust, are you using maven mirror or proxy? settings.xml would be useful

Could you try building this branch
https://github.com/os72/protoc-jar/tree/support_mirror

And run
`java -jar protoc-jar-3.7.0.1-SNAPSHOT.jar --version -v351`

**Comment by cjjnjust** [comment]
I 'm using maven behind a proxy and maven works well, settings.xml is correctly set.

I tried build the branch and following is output for the command:

java -jar protoc-jar-3.7.0.1-SNAPSHOT.jar --version -v351
`
protoc-jar: protoc version: 3.5.1, detected platform: linux-x86_64 (linux/amd64)
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/github/os72/protoc/maven-metadata.xml
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/github/os72/protoc/maven-metadata.xml
protoc-jar: java.io.FileNotFoundException: Unsupported platform: protoc-3.5.1-linux-x86_64.exe
java.io.FileNotFoundException: Unsupported platform: protoc-3.5.1-linux-x86_64.exe
	at com.github.os72.protocjar.Protoc.extractProtoc(Protoc.java:231)
	at com.github.os72.protocjar.Protoc.extractProtoc(Protoc.java:192)
	at com.github.os72.protocjar.Protoc.runProtoc(Protoc.java:76)
	at com.github.os72.protocjar.Protoc.runProtoc(Protoc.java:63)
	at com.github.os72.protocjar.Protoc.main(Protoc.java:53)
`

**Comment by os72** [comment]
So you have a `proxies` section, not `mirrors`? Do you mind sharing it? (remove any sensitive info of course)

**Comment by cjjnjust** [comment]
This is proxies section from my local settings.xml.    
<proxies>
        <proxy>
            <id>http-proxy</id>
            <active>true</active>
            <protocol>http</protocol>
            <host>web-proxy.oa.com</host>
            <port>8080</port>
    </proxy>
        <proxy>
            <id>https-proxy</id>
            <active>true</active>
            <protocol>https</protocol>
            <host>web-proxy.oa.com</host>
            <port>8080</port>
        </proxy>

</proxies>

**Comment by os72** [comment]
Thanks

**Comment by os72** [comment]
Hi @cjjnjust, could you pull the changes and try again? Thanks

**Comment by cjjnjust** [comment]
Hi @os72 
It works!  Thanks.

`
java -jar protoc-jar-3.7.0.1-SNAPSHOT.jar --version -v351
protoc-jar: protoc version: 3.5.1, detected platform: linux-x86_64 (linux/amd64)
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/maven-metadata.xml [HTTP @ web-proxy.oa.com/10.24.32.55:8080]
protoc-jar: saved: /tmp/protocjar.webcache/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/github/os72/protoc/maven-metadata.xml [HTTP @ web-proxy.oa.com/10.24.32.55:8080]
protoc-jar: saved: /tmp/protocjar.webcache/com/github/os72/protoc/maven-metadata.xml
protoc-jar: cached: /tmp/protocjar.webcache/com/google/protobuf/protoc/maven-metadata.xml
protoc-jar: downloading: https://repo.maven.apache.org/maven2/com/google/protobuf/protoc/3.5.1/protoc-3.5.1-linux-x86_64.exe [HTTP @ web-proxy.oa.com/10.24.32.55:8080]
protoc-jar: saved: /tmp/protocjar.webcache/com/google/protobuf/protoc/3.5.1/protoc-3.5.1-linux-x86_64.exe
protoc-jar: executing: [/tmp/protocjar118196032159410809/bin/protoc.exe, --version]
libprotoc 3.5.1

`

**Comment by os72** [comment]
Great! Will release soon

**Comment by cjjnjust** [comment]
Hi @os72, 

Will this get release in this week? Do we have release plan? Thanks.

**Comment by os72** [comment]
Yes this week. Wanted to include one more feature

**Comment by os72** [comment]
Released in version 3.7.0.1
